### PR TITLE
Add apply_join mode to inner, left and semi joins

### PR DESCRIPTION
**Pull request by jwimberl**
* Adds apply_join mode to these three meta join modes
* Adds test cases, temporarily skipping the check that the semi-join output has the same schema as the reference due to https://github.com/Paradigm4/equi_join/issues/27

### CODE REVIEW

**Comment by jwimberl** [comment]
The unit tests pass locally on a SciDB 23 machine with the exception of several failures related to SciDB data frames (issue #44) and datetime (issue #46)

**Comment by jwimberl** [comment]
The unit tests also pass locally on a SciDB 19.11 machine, in which case there are no errors at all (issues #44 and #46 being specific to SciDB 23.10)

File: tests/testthat/test_join.R [hunk 1]
```diff
@@ -15,9 +15,10 @@ test_that("join with no conflicting field names", {
   rightDf = data.frame(rda = 3:10, rdb = 13:20, rfa = LETTERS[3:10], rfb = 3.14 * 3:10)
   leftTemplate = conn$array_from_schema("<lfa:string, lfb:double> [lda;ldb] ")
   rightTemplate = conn$array_from_schema("<rfa:string, rfb:double> [rda;rdb] ")
-  
-  L = conn$array_from_df(leftDf, leftTemplate, force_template_schema = T)
-  R = conn$array_from_df(rightDf, rightTemplate, force_template_schema = T)
+  # At least one input to 'apply_join' must have a specified chunk size.
+  # So we need to persist the arrays
+  L = conn$array_from_df(leftDf, leftTemplate, force_template_schema = T)$change_schema(leftTemplate)$persist(.gc = F)
```
Comment (smoore-p4) [inline tests/testthat/test_join.R:20]: Is the `.gc = F` required here? I'm a little concerned about leaving arrays hanging around in general

**Review by smoore-p4 — CHANGES_REQUESTED** [review]
Just the bits on the test using `$persist(.gc=False)` and/or add in the `$remove_array()` for them

File: tests/testthat/test_join.R (hunk 1 above)
Comment (jwimberl) [reply tests/testthat/test_join.R:20]: While the `.gc = F` may not be strictly necessary, persisting the array is required for some of the join cases to work. I've uncommented the `remove_array()` calls for this block in a new commit.

File: tests/testthat/test_join.R [hunk 2]
```diff
@@ -173,10 +211,12 @@ test_that("join with no conflicting field names", {
...
   test_inner_join()
+  test_apply_inner_join()
   test_select_fields()
   test_select_fields_dry()
   test_join_settings()
   test_left_join()
+  test_apply_left_join()
   test_right_join()
   
   # L$remove_array()
```
[the remark this answers is not in the document]
Comment (jwimberl) [reply tests/testthat/test_join.R]: Changed in a new commit

**Review by smoore-p4 — APPROVED** [review]

**Pull request merged**